### PR TITLE
(fix) revised kucoin connector from exchange_data_fetcher

### PR DESCRIPTION
**Pull request by vic-en**
**Before submitting this PR, please make sure**:

- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**After creating this PR, please make sure:**

- [ ] You link the related GitHub issue or ticket

**A description of the changes proposed in the pull request**:
This is the kucoin connector from the exchange_data_fetcher. It will need some extra work to be compatible with the current structure of hummingbot

### CODE REVIEW

**Review by vic-en — COMMENTED** [review]
LGTM

**Review by Nullably — APPROVED** [review]
LGTM
One minor thing on line 374 to 376, should you use the new `sec_until_next_oclock` function for this?
https://github.com/CoinAlpha/hummingbot/pull/2507/files#diff-93769022a39687aca6c9a05e6b52de552dc82c1b4306a9761438c5103896d22cR374

File: hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py [hunk 1]
```diff
@@ -145,74 +161,194 @@ async def ws_connect_data(self):
...
                     ws: websockets.WebSocketClientProtocol = ws
-                    for trading_pair in self._trading_pairs:
-                        subscribe_request: Dict[str, Any] = {
-                            "id": int(time.time()),
-                            "type": "subscribe",
-                            "topic": f"/market/level2:{trading_pair}",
-                            "response": True
-                        }
-                        await ws.send(json.dumps(subscribe_request))
+                    ping_task: asyncio.Task = \
```
Comment (martinkou) [inline hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py:250]: `ping_task` is used in `finally`, but its definition here depends on a successful `websockets.connect()`.

This is not good, and could lead to `_outer_messages()` exiting silently if `websockets.connect()` happens to fail.

What you should do instead, is to define `ping_task` as optional and initialize as None before the try statement. And in the finally block, check if `ping_task` is defined and cancel it there.

**Review by martinkou — CHANGES_REQUESTED** [review]

File: hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py [hunk 2]
```diff
@@ -145,74 +161,194 @@ async def ws_connect_data(self):
...
+        return market_subsets
+
+    async def _start_update_tasks(self, stream_type: StreamType, output: asyncio.Queue):
+        self._stop_update_tasks(stream_type)
+        market_assignments = await self.get_markets_per_ws_connection()
+
+        task_dict: dict = {}
+        for task_index, market_subset in enumerate(market_assignments):
+            task_dict[task_index] = {"markets": set(market_subset.split(',')),
+                                     "task": asyncio.ensure_future(self._outer_messages(stream_type,
```
Comment (martinkou) [inline hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py:199]: All background tasks in hummingbot must use `safe_ensure_future()`, to properly log any background tasks that failed silently.

File: hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py [hunk 3]
```diff
@@ -145,74 +161,194 @@ async def ws_connect_data(self):
...
-                    for trading_pair in self._trading_pairs:
-                        subscribe_request: Dict[str, Any] = {
-                            "id": int(time.time()),
-                            "type": "subscribe",
-                            "topic": f"/market/level2:{trading_pair}",
-                            "response": True
-                        }
-                        await ws.send(json.dumps(subscribe_request))
+                    ping_task: asyncio.Task = \
+                        asyncio.ensure_future(self._send_ping(KucoinAPIOrderBookDataSource.PING_INTERVAL, ws))
```
Comment (martinkou) [inline hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py:251]: Same here, use `safe_ensure_future()`

**Review by martinkou — APPROVED** [review]

**Pull request merged**